File: app/src/main/java/com/eli/nationalid/Activities/MainActivity.java
package com.eli.nationalid.Activities;

import android.app.PendingIntent;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ImageDecoder;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.browser.customtabs.CustomTabsIntent.Builder;
import androidx.core.content.res.ResourcesCompat;

import com.eli.nationalid.R;
import com.eli.nationalid.util.BaseActivity;
import com.eli.nationalid.util.Constants;

import java.io.IOException;
import java.io.InputStream;

public class MainActivity extends BaseActivity {
    private  Bitmap icon;
    private    ImageDecoder.Source source;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);


    }

    @Override
    protected void onStart() {
        super.onStart();
        /** Use a CustomTabsIntent.Builder to configure CustomTabsIntent.Once ready,
         * call CustomTabsIntent.Builder.build() to create a CustomTabsIntentand
         * launch the desired Url with CustomTabsIntent.launchUrl()
         **/
        openCustomTab();

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    @Override
    protected void onRestart() {
        super.onRestart();
    }

    public void openCustomTab() {




        Builder builder = new CustomTabsIntent.Builder();
        builder.setToolbarColor(getResources().getColor(R.color.secondaryDarkColor));
        builder.setShowTitle(true);
        builder.addDefaultShareMenuItem();

        Intent actionIntent = new Intent(this.getApplicationContext(), ShareBroadcastReceiver.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(getApplicationContext(), 0, actionIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        icon = BitmapFactory.decodeResource(this.getResources(), android.R.drawable.ic_menu_share);

        /**
         * Create {@link PendingIntent} and another {@link CustomTabsIntent}
         * To open nuxtjs website
         * **/
        CustomTabsIntent anotherCustomTab = new CustomTabsIntent.Builder()
                .setToolbarColor(getResources().getColor(R.color.secondaryDarkColor))
                .addDefaultShareMenuItem()
                .setShowTitle(true).build();
        int requestCode = 100;
        Intent intent = anotherCustomTab.intent;
        intent.setData(Uri.parse(Constants.NUXT_URL));

        /**
         * Vuetify Intents
         * **/
        CustomTabsIntent vueTifyCustomTab = new CustomTabsIntent.Builder()
                .setToolbarColor(getResources().getColor(R.color.secondaryDarkColor))
                .addDefaultShareMenuItem()
                .setShowTitle(true).build();
        Intent vueTifyIntent = vueTifyCustomTab.intent;
        vueTifyIntent.setData(Uri.parse(Constants.VUETIFY_URL));

        /**
         * Composition Api Intent
         * **/
        CustomTabsIntent compositionCustomTab = new CustomTabsIntent.Builder()
                .setToolbarColor(getResources().getColor(R.color.secondaryDarkColor))
                .addDefaultShareMenuItem()
                .setShowTitle(true).build();
        Intent compositionIntent = compositionCustomTab.intent;
        compositionIntent.setData(Uri.parse(Constants.COMPOSITION_API_URL));

        PendingIntent pendingIntent1 = PendingIntent.getActivity(this,requestCode,intent, PendingIntent.FLAG_CANCEL_CURRENT);
        PendingIntent VUETIFY_INTENT = PendingIntent.getActivity(this,requestCode,vueTifyIntent,PendingIntent.FLAG_CANCEL_CURRENT);
        PendingIntent COMPOSITION_API_INTENT = PendingIntent.getActivity(this, requestCode,compositionIntent, PendingIntent.FLAG_CANCEL_CURRENT);

        builder.setActionButton(icon, "Share ", pendingIntent);
        builder.addMenuItem("NuxitJs", pendingIntent1);
        builder.addMenuItem("VuetifyJs",VUETIFY_INTENT);
        builder.addMenuItem("Composition API", COMPOSITION_API_INTENT);
        CustomTabsIntent customTabsIntent = builder.build();
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        customTabsIntent.launchUrl(this, Uri.parse(Constants.VUEJS_URL));

    }


}

File: app/src/main/java/com/eli/nationalid/Activities/MainActivity.kt
package com.eli.nationalid.Activities

import android.app.AlertDialog
import android.app.PendingIntent
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.view.GravityCompat
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.eli.nationalid.Api.ApiClient
import com.eli.nationalid.R
import com.eli.nationalid.adapters.RecyclerAdapter
import com.eli.nationalid.models.DataModule
import com.eli.nationalid.room.database.DatabaseClient
import com.eli.nationalid.room.entities.DataListEntity
import com.eli.nationalid.room.model.infoListViewModel
import com.eli.nationalid.ui.viewmodel.DataListViewModel
import com.eli.nationalid.util.*
import com.google.android.material.navigation.NavigationView
import com.student.Api.ApiInterface
import kotlinx.android.synthetic.main.main_activity.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : BaseActivity() {
    var model: DataListViewModel? = null
    var databaseModel: infoListViewModel? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
        configureToolBar();
        configureNavigationDrawer();
        model = ViewModelProviders.of(this).get(DataListViewModel::class.java)
        databaseModel = ViewModelProviders.of(this).get(infoListViewModel::class.java)

    }


    private fun configureToolBar() {
        setSupportActionBar(toolbar)
        toolbar.setNavigationIcon(R.drawable.ic_menu_black_24dp)
        toolbar.inflateMenu(R.menu.navigation)
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        supportActionBar?.setDisplayShowTitleEnabled(false)
    }


    override fun onResume() {
        super.onResume()

        databaseModel?.itemInfoList?.observe(this, Observer<List<DataListEntity?>?> { t ->
            if(t.isNullOrEmpty()){
                callRemoteDataAndSaveToLocalDatabase()
            }
            createList(t as List<DataListEntity>)
        })

       // val data = model?.getDataList()

    }
    private fun configureNavigationDrawer() {
        navigation.setNavigationItemSelectedListener(object : NavigationView.OnNavigationItemSelectedListener {
            override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
                val itemId: Int = menuItem.getItemId()
                if (itemId == R.id.nav_about_us) {
                    startActivity(Intent(this@MainActivity, AboutUsActivity::class.java))
                } else if (itemId == R.id.nav_twitter) {
                   openTwitter();
                } else if (itemId == R.id.nav_facebook) {
                   openFacebook()
                } else if (itemId == R.id.nav_linkedin) {
                    openLinkedin()
                } else if (itemId == R.id.nav_youtube) {
                   openYoutubeChannel()
                } else if (itemId == R.id.nav_contact_us) {
                    //startActivity(Intent(this@MainActivity, ContactUsActivity::class.java))
                } else if (itemId == R.id.nav_general_info) {

                } else if (itemId == R.id.nav_home) {
                    toggle()

                } else if (itemId == R.id.nav_our_site) {

                }
                return false
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if(item.getItemId() == android.R.id.home){ toggle() }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        return true
    }

    private fun toggle() {
        if (drawerLayout.isDrawerVisible(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START)
        } else { drawerLayout.openDrawer(GravityCompat.START) }
    }

    private fun createList(apartmentData: List<DataListEntity>) {
        recyclerview.layoutManager = LinearLayoutManager(this)
        recyclerview.isNestedScrollingEnabled = false
        recyclerview.setHasFixedSize(true)
        recyclerview.setItemViewCacheSize(20)
        recyclerview!!.adapter = RecyclerAdapter(this, apartmentData)
        recyclerview!!.visibility = View.VISIBLE
    }

    fun openCustomTab(TAB_URL: String?) {
        val builder = CustomTabsIntent.Builder()
        //builder.setToolbarColor(resources.getColor(R.color.secondaryDarkColor, null))
        builder.setShowTitle(true)
        builder.addDefaultShareMenuItem()
        val actionIntent = Intent(this.applicationContext, ShareBroadcastReceiver::class.java)
        val pendingIntent = PendingIntent.getBroadcast(applicationContext, 0, actionIntent, PendingIntent.FLAG_UPDATE_CURRENT)

        builder.setActionButton(BitmapFactory.decodeResource(this.resources, android.R.drawable.ic_menu_share), "Share ", pendingIntent)
        val customTabsIntent = builder.build()
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        customTabsIntent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        customTabsIntent.launchUrl(this, Uri.parse(TAB_URL))
    }



    override fun onBackPressed() {
        AlertDialog.Builder(this)
                .setMessage("Are you sure you want to exit this app ?")
                .setCancelable(false)
                .setPositiveButton("YEs"){ _, _ -> super.onBackPressed() }
                .setNegativeButton("No", null)
                .show()

    }

    override fun onPause() {
        super.onPause()
        callRemoteDataAndSaveToLocalDatabase()

    }

    fun callRemoteDataAndSaveToLocalDatabase(){
        var apiInterface = ApiClient.getInstance().create(ApiInterface::class.java)
        val call = apiInterface.getDataList()
        call.enqueue(object : Callback<List<DataModule>?> {
            override fun onResponse(call: Call<List<DataModule>?>, response: Response<List<DataModule>?>) {
                //val statusCode = response.code()
                val user: List<DataModule>? = response.body()
                if (user != null) {
                    for (item in user){
                        var datalistItem = DataListEntity(item.mId!!,item.mTitle!!,item.mSubtitle!!,item.mIcon!!,item.mUrlink!!);
                        DatabaseClient.getmInstance(baseContext).getAppDatabase()
                                .infoModelDao()
                                .insertDataList(datalistItem)
                    }
                }


            }

            override fun onFailure(call: Call<List<DataModule>?>, t: Throwable) {
                // Log error here since request failed
            }
        })
    }
}


